Some Python bindings fixes, and raise test coverage to 82%

bowlofeggs opened the pull request:
This pull request contains 6 commits. The first 4 commits fix various problems discovered while testing ```bodhi.client.bindings```, the 5th commit adds many tests for the bindings, and the last commit requires test coverage to remain at 82%.

bowlofeggs committed:
Remove an unusable delete() method from the bindings.

bodhi.client.bindings.BodhiClient.delete() could not have been
working since it uses @errorhandled and does not return any
response. The errorhandled() decorator will fail when any decorated
method returns None. Since the delete() method contained a warning
that it was deprecated, it was determined that it was better to go
ahead and delete it rather than fix it.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Fix a query regular expression so that Fedora update ids work.

This commit allows bodhi.client.bindings.BodhiClient.query to
accept update IDs in the form of FEDORA-2017-XXXX by fixing a
regular expression.

This problem was originally reported and fixed by Adam Williamson
in fedora.client.bodhi, but the same code existed in
bodhi.client.bindings.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Correct the docblock for BodhiClient.list_overrides.

bodhi.client.bindings.BodhiClient.list_overrides() had a docblock
that was copied from save_overrides(). This commit corrects the
docblock to say that the method is for listing overrides.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Convert a method into a staticmethod, since self is not used.

bodhi.client.bindings.BodhiClient.override_str() does not use self.
This commit converts that method into a staticmethod so it can be
used outside of a BodhiClient instance.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Greatly increase the test coverage on the Python bindings.

This commit adds several tests for bodhi.client.bindings, raising
the test coverage to 82%.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Require test coverage to meet or exceed 82%.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

AdamWill commented:
For the record, the regex bug was not only "originally reported and fixed" by me, it was *created*, reported and fixed by me. :P

bowlofeggs commented:
@AdamWill hahahah well I'm still happy to have the fix ☺

jeremycline commented on bodhi/client/bindings.py:
```diff
@@ -230,7 +222,7 @@ def query(self, **kwargs):
             if re.search(r'(\.el|\.fc)\d\d?', kwargs['package']):
                 kwargs['builds'] = kwargs['package']
-            elif re.search(r'FEDORA-(EPEL)?-\d{4,4}', kwargs['package']):
+            elif re.search(r'FEDORA-(EPEL-)?\d{4,4}', kwargs['package']):
```
Hmm, what happens in the year 10,000? Think of the children!

bowlofeggs replied:
Hahahaha, well I guess we'll have to go back to the drawing board, but only if humans still exist then…

jeremycline commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -24,6 +25,435 @@
+        client.csrf_token = 'a token'
+        client.send_request = mock.MagicMock(return_value='response')
+
+        response = client.comment('bodhi-2.4.0-1.fc25', 'It ate my cat!', karma=-1, email=True)
```
I hate when that happens 🐱

bowlofeggs replied:
☺

jeremycline commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -24,6 +25,435 @@
+
+        self.assertEqual(override, 'this is an override')
+
+    def test_with_str(self):
```
I think you got these mixed up, this one uses a dict, the other one uses a string

jeremycline commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -24,6 +25,435 @@
+
+        self.assertEqual(result, 'return_value')
+        client.send_request.assert_called_once_with(
+            'updates/', verb='GET', params={'packages': 'bodhi', 'releases': 'f26'})
```
Should it be wrapping the value of 'releases' in a list if it's not already a list? I just noticed the docblock says it can be a list (and its name implies that) so I wondered.

bowlofeggs replied:
The ```release``` argument is actually there for backwards compatibility with the bodhi 1 client. I'm not sure anything would be using this, especially with the newly relocated bindings. However, I decided to keep it for now to minimize changes. Perhaps it can be removed with Bodhi 3.

I think you are right that it should be converting it to a list. I'll add another commit that fixes this and modifies the test.

jeremycline commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -24,6 +25,435 @@
+        response = client.save(**kwargs)
+
+        self.assertEqual(response, 'return_value')
+        kwargs['type'] = kwargs['type_']
```
Should it be popping the old ``type_`` kwarg before sending the request?

bowlofeggs replied:
The request got sent two lines above in the call to ```client.save()```. This line is just adjusting the kwargs for the purpose of asserting that ```save()``` properly modified the kwargs.

jeremycline replied:
I think what I meant here was, should it be asserting that both ``type`` and ``type_`` are keywords in the data dictionary that ``send_request`` is called with?

jeremycline commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -24,6 +25,435 @@
+        # date sent is within 2 seconds of the now variable. It would be surprising if it took more
+        # than 2 seconds to start the function and execute its first instruction!
+        expected_expiration = now + timedelta(days=2)
+        self.assertTrue((actual_expiration - expected_expiration) < timedelta(seconds=2))
```
This may be a flaky test. Not because it's likely it'll take more than two seconds to run (although I guess that's possibly with very unlucky scheduling), but because if a system clock update occurs between the first call and the second call to ``utcnow`` and the drift was more than 2 seconds (which seems like a very small number for drift) this won't be true.

Maybe that won't happen much (or ever!), just thought I'd bring it up.

bowlofeggs replied:
Do you have any specific recommendations for what you think would be better here? I think those scenarios are fairly unlikely, though not impossible. If they were to happen, we could raise the threshold later.

jeremycline replied:
Not really, unfortunately. You could probably raise it to something like 5 minutes if the expected expiration is two days. Two days is 2880 minutes so give or take 5 doesn't seem like a big deal. It's totally up to you.

jeremycline reviewed (changes requested):
Overall it looks good, there's just a pair of tests whose names need switching. There's also a time-based test which you may want to re-consider, but I leave that up to your judgment.

bowlofeggs committed:
Wrap the release kwarg in a list if it isn't on in query().

This commit alters bodhi.client.bindings.BodhiClient.query() such
that if the release kwargs is not a list it will convert it to a
list before passing it to releases.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs commented:
@jeremycline I've rebased this pull request and fixed everything you noted except for the timing issue and the kwargs popping (since that was for test assertion). What do you think?

jeremycline commented:
I clarified my comments, but neither is a blocking issue to me, so merge whenever you're ready :shipit:

bowlofeggs commented:
On Tue, 2017-03-07 at 11:03 -0800, Jeremy Cline wrote:
> I think what I meant here was, should it be asserting that both type
> and type_ are keywords in the data dictionary that send_request is
> called with?

That is what the underlying library is doing, and this is again to keep
compatibility with the v1 library. Since I have no knowledge of Bodhi
v1 it's a little hard to say what should happen here, though it does
seem that dropping the extra kwarg makes sense. I'm inclined to leave
it as-is for now and just drop all this code with Bodhi 3.

Is that OK by you?

bowlofeggs commented:
Ah I will take your last comment to indicate that it is OK by you. Thanks for the review! I'll fix that window to be 5 minutes.
